pkg/scheduler/util/scheduler_helper.go

```diff
@@ -55,53 +60,125 @@ func PredicateNodes(task *api.TaskInfo, nodes []*api.NodeInfo, fn api.PredicateF
+	priorityConfigs []algorithm.PriorityConfig,
+) (schedulerapi.HostPriorityList, error) {
+	nodeNameToInfo, nodes := generateNodeMapAndSlice(filterNodes)
+	var (
+		mu   = sync.Mutex{}
+		wg   = sync.WaitGroup{}
+		errs []error
+	)
+	appendError := func(err error) {
+		mu.Lock()
+		defer mu.Unlock()
+		errs = append(errs, err)
+	}
 
-	var workerLock sync.Mutex
-	scoreNode := func(index int) {
-		node := nodes[index]
-		score, err := fn(task, node)
-		if err != nil {
-			glog.Errorf("Error in Calculating Priority for the node:%v", err)
-			return
+	results := make([]schedulerapi.HostPriorityList, len(priorityConfigs), len(priorityConfigs))
+
+	for i, priorityConfig := range priorityConfigs {
+		if priorityConfig.Function != nil {
+			wg.Add(1)
+			go func(index int, config algorithm.PriorityConfig) {
+				defer wg.Done()
+				var err error
+				results[index], err = config.Function(task.Pod, nodeNameToInfo, nodes)
+				if err != nil {
+					appendError(err)
+				}
+			}(i, priorityConfig)
+		} else {
+			results[i] = make(schedulerapi.HostPriorityList, len(nodes))
+		}
+	}
+	processNode := func(index int) {
+		nodeInfo := nodeNameToInfo[nodes[index].Name]
+		var err error
+		for i := range priorityConfigs {
+			if priorityConfigs[i].Function != nil {
+				continue
+			}
+			results[i][index], err = priorityConfigs[i].Map(task.Pod, nil, nodeInfo)
+			if err != nil {
+				appendError(err)
+				return
+			}
 		}
+	}
+	workqueue.ParallelizeUntil(context.TODO(), 16, len(nodes), processNode)
+	for i, priorityConfig := range priorityConfigs {
+		if priorityConfig.Reduce == nil {
+			continue
+		}
+		wg.Add(1)
+		go func(index int, config algorithm.PriorityConfig) {
+			defer wg.Done()
+			if err := config.Reduce(task.Pod, nil, nodeNameToInfo, results[index]); err != nil {
+				appendError(err)
+			}
+			if glog.V(10) {
+				for _, hostPriority := range results[index] {
+					glog.Infof("%v -> %v: %v, Score: (%d)", task.Pod.Name, hostPriority.Host, config.Name, hostPriority.Score)
+				}
+			}
+		}(i, priorityConfig)
+	}
+	// Wait for all computations to be finished.
+	wg.Wait()
+	if len(errs) != 0 {
+		return schedulerapi.HostPriorityList{}, errors.NewAggregate(errs)
+	}
 
-		workerLock.Lock()
-		nodeScores[score] = append(nodeScores[score], node)
-		workerLock.Unlock()
+	// Summarize all scores.
+	result := make(schedulerapi.HostPriorityList, 0, len(nodes))
+	for i := range nodes {
+		result = append(result, schedulerapi.HostPriority{Host: nodes[i].Name, Score: 0})
+		for j := range priorityConfigs {
+			result[i].Score += results[j][i].Score * priorityConfigs[j].Weight
+		}
 	}
-	workqueue.ParallelizeUntil(context.TODO(), 16, len(nodes), scoreNode)
-	return nodeScores
+
+	return result, nil
 }
 
 // SortNodes returns nodes by order of score
-func SortNodes(nodeScores map[float64][]*api.NodeInfo) []*api.NodeInfo {
+func SortNodes(priorityList schedulerapi.HostPriorityList, nodesInfo map[string]*api.NodeInfo) []*api.NodeInfo {
```

Comment:
Score's type become `int` now. It will revert https://github.com/kubernetes-sigs/kube-batch/pull/719, does it matter? @k82cn @lmzqwer2

Reply:
@hex108 @thandayuthapani The score calculated by priorities in k8s are all return int type now and  we do not need change it to float now.  In the future, if  priorities in k8s change the return value type, we can update correspondingly

Reply:
The details is in https://github.com/kubernetes-sigs/kube-batch/issues/708.

Reply:
yes, I have saw that, but for now we have no function that return float type value. and for new function, I think it may have little affect on the result because we have many  priorities and every has its weight.  And we use some struct such as `HostPriority` from k8s. If we want to use float type, I think we should first change the type in k8s. So I think int type is enough for now. @k82cn @thandayuthapani @hex108

Reply:
We need to keep float here, and float also support int for k8s's default priorities.

Reply:
@k82cn @hex108 @thandayuthapani Updated, PTAL